### PR TITLE
XGLM: Fix left-padding (PT and TF)

### LINKED ISSUE #22707: the generated results are different between generating a batch input_ids and a single sequence input_ids
### System Info

- `transformers` version: 4.28.0.dev0
- Platform: Linux-4.15.0-142-generic-x86_64-with-glibc2.23
- Python version: 3.9.16
- Huggingface_hub version: 0.13.4
- Safetensors version: not installed
- PyTorch version (GPU?): 2.0.0+cu117 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: <fill in>
- Using distributed or parallel set-up in script?: <fill in>

### Who can help?

@ArthurZucker @gante @youne

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

I finetuned a pre-trained XGLM model with LoRA, and when I use the model to generate sequences with a same input ( the only diffience between two ways is one sentence is in a list with other sentence and generate them together, the other is a single sentence in a list )，the result is different with the same setting!
the details are following:

tokenizer = XGLMTokenizer.from_pretrained(base_model)
tokenizer.padding_side = "left"
tokenizer.add_special_tokens({'additional_special_tokens': ['=']})
model = XGLMForCausalLM.from_pretrained(base_model,device_map="auto",)
model = PeftModel.from_pretrained(model,lora_weights,)
model.eval()
model = torch.compile(model)
p = ['今年 , 这种 主导作用 依然 非常 突出 。 = ', '国际足联 将 严惩 足球场 上 的 欺骗 行为 = ', '枪手 被 警方 击毙 。 = ']
a = tokenizer(p, padding=True, return_tensors='pt')
ga = model.generate(**a, num_beams=5, max_new_tokens=128)
oa = tokenizer.batch_decode(ga, skip_special_tokens=False)
b = ['今年 , 这种 主导作用 依然 非常 突出 。 = ']  # b equals p[0]
b = tokenizer(b, padding=True, return_tensors='pt')
gb = model.generate(**b, num_beams=5, max_new_tokens=128)
ob = tokenizer.batch_decode(gb, skip_special_tokens=False)

result:
# print(oa)
['<pad><pad><pad><pad><pad></s> 今年, 这种 主导作用 依然 非常 突出 。= this year, this dominant role is still very prominent.</s>', '</s> 国际足联 将 严惩 足球场 上 的 欺骗 行为= fifa to punish cheaters in stadiums</s><pad><pad><pad><pad><pad>', '<pad><pad><pad><pad><pad><pad><pad><pad><pad></s> 枪手 被 警方 击<unk> 。= the gunman was killed by police.</s><pad><pad><pad><pad>']
# print(ob)
['</s> 今年, 这种 主导作用 依然 非常 突出 。= this year, this dominant role has still been very prominent.</s>']

the output following the "=" token is different in the case.

### Expected behavior

the output following the "=" token is different in the case.
I am confused why the same input but different output just because of the additional <pad> token? But the <pad> attention mask is 0. 
And I want to know how to change the code to make them be same?
Is it a bug or is there a problem with the code?
Looking forward to a quick response, and thanks very much!

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

Fixes left-padding for XGLM, on PT and TF. It is the usual problem: `position_ids` was not being used/passed around, and the code assumed that position ids = past length + input length (which is not true when left-padding is present).

Fixes #22707 

While touching XGLM, other issues were sorted:
1. on PT, docs were duplicated (and one of the copies was wrong)
2. XGLM generate with sampling integration test was pinned on CPU (as always, GPU gives different results, which was making our slow CI report an error)
3. TF XLA was failing because of this (left padding support), so now we have TF XLA XGLM 🙌

### CODE REVIEW

File: src/transformers/models/xglm/modeling_xglm.py
Comment (gante) [inline src/transformers/models/xglm/modeling_xglm.py:78]: `XGLM_INPUTS_DOCSTRING` did not contain the actual arguments of `forward`, and `forward` had a correct docstring (which meant that the docs had a two sets of inputs 👀 )

This PR fixes that

File: tests/models/xglm/test_modeling_xglm.py
Comment (gante) [inline tests/models/xglm/test_modeling_xglm.py:410]: Two notes here:
1. This output did not change from the previous version, despite the left-padding being much bigger due to the other sentence in the batch;
2. On `main`, this modified test fails -- this new test protects us from regressions 👍

**Comment by gante** [comment]
@amyeroberts now it is ready :)

File: src/transformers/models/xglm/modeling_flax_xglm.py
Comment (gante) [inline src/transformers/models/xglm/modeling_flax_xglm.py:127]: (unused stand-alone method, probably copy-paste; couldn't find any import)

File: src/transformers/models/xglm/modeling_tf_xglm.py
Comment (gante) [inline src/transformers/models/xglm/modeling_tf_xglm.py:479]: (the fixes here are pretty much the same as in the PT side)

File: tests/models/xglm/test_modeling_tf_xglm.py
Comment (gante) [inline tests/models/xglm/test_modeling_tf_xglm.py:178]: (the changes here are pretty much the same as on the PT side)

File: tests/models/xglm/test_modeling_xglm.py
Comment (amyeroberts) [reply tests/models/xglm/test_modeling_xglm.py:410]: Just to double check, for 2. this test is failing but it runs successfully here?

Comment (gante) [reply tests/models/xglm/test_modeling_xglm.py:410]: @amyeroberts the original test is passing on `main`. This modified test would fail on `main`, but runs successfully after the modelling changes included in the PR.

File: src/transformers/models/xglm/modeling_xglm.py
Comment (amyeroberts) [inline src/transformers/models/xglm/modeling_xglm.py:868]: I'm a bit confused 😅  Why do we set `position_ids` as `None` if they're not `None` when passed in?

**Review by amyeroberts — APPROVED** [review]
Thanks for the fix and verification with the tests! Changes LGTM - good to have to docstrings up-to-date too 🙏 

Asking for a second opinion from @sgugger before merging on the change of `forward` signature for the `XGLMForCausalLM` model. For the other layers, I think it's fine, but as this is quite an old model, this might be too much of a breaking change.

**Comment by sgugger** [comment]
Adding a new keyword argument is not considered breaking, so that's fine!

File: src/transformers/models/xglm/modeling_xglm.py
Comment (gante) [reply src/transformers/models/xglm/modeling_xglm.py:868]: More than a valid question, this is a copy-paste from other models that (IMO) should be changed!

`position_ids` should be computed from `attention_mask`, so the `else` path invalidates them when `attention_mask is None`. However, that doesn't explain the `and position_ids is None` in the `if` 🙃 

A better change would be to move this `position_ids` computation to the modeling code itself, for all models. However, I'm not entirely sure whether researchers would miss being able to control `position_ids`.

**Pull request merged**
